crates/matrix-sdk/src/encryption/backups/mod.rs

```diff
@@ -0,0 +1,1157 @@
+        }
+
+        olm_machine.store().delete_secrets_from_inbox(&SecretName::RecoveryKey).await?;
+
+        Ok(())
+    }
+
+    /// Check and re-enable a backup if we have a backup recovery key locally.
+    async fn maybe_resume_backups(&self) -> Result<(), Error> {
+        let olm_machine = self.client.olm_machine().await;
+        let olm_machine = olm_machine.as_ref().ok_or(Error::NoOlmMachine)?;
+
+        // Let us first check if we have a stored backup recovery key and a backup
+        // version.
+        if !self.resume_backup_from_stored_backup_key(olm_machine).await? {
+            // We didn't manage to enable backups from a stored backup recovery key, let us
+            // check our secret inbox. Perhaps we can find a valid key there.
+            self.maybe_resume_from_secret_inbox(olm_machine).await?;
+        }
+
+        Ok(())
+    }
+
+    /// Listen for `m.secret.send` to-device events and check the secret inbox
+    /// if we do receive one.
+    pub(crate) async fn secret_send_event_handler(_: ToDeviceSecretSendEvent, client: Client) {
+        let olm_machine = client.olm_machine().await;
+
+        // TODO: Because of our crude multi-process support, which reloads the whole
+        // [`OlmMachine`] the `secrets_stream` might stop giving you updates. Once
+        // that's fixed, stop listening to individual secret send events and
+        // listen to the secrets stream.
+        if let Some(olm_machine) = olm_machine.as_ref() {
+            if let Err(e) =
+                client.encryption().backups().maybe_resume_from_secret_inbox(olm_machine).await
+            {
+                error!("Could not handle `m.secret.send` event: {e:?}");
+            }
+        } else {
+            error!("Tried to handle a `m.secret.send` event but no OlmMachine was initialized");
+        }
+    }
+
+    /// Try to enable backups using the secret found in the [`GossippedSecret`].
+    ///
+    /// The [`GossippedSecret`] represents a `m.secret.send` event we received
+    /// from a trusted device and have pulled out of the secret inbox.
+    ///
+    /// Returns true if we enabled backups, false otherwise.
+    pub(crate) async fn handle_received_secret(
+        &self,
+        secret: GossippedSecret,
+    ) -> Result<bool, Error> {
+        if secret.secret_name == SecretName::RecoveryKey {
+            if self.maybe_enable_backups(&secret.event.content.secret).await? {
+                let olm_machine = self.client.olm_machine().await;
+                let olm_machine = olm_machine.as_ref().ok_or(Error::NoOlmMachine)?;
+                olm_machine.store().delete_secrets_from_inbox(&secret.secret_name).await?;
+
+                Ok(true)
+            } else {
+                Ok(false)
+            }
+        } else {
+            Ok(false)
+        }
+    }
+
+    /// Send a notification to the task which is responsible to upload room keys
+    /// to the backup that it might have new room keys to back up.
```

Comment:
```suggestion
    /// Send a notification to the task which is responsible for uploading room
    /// keys to the backup that it might have new room keys to back up.
```